exercises-cpp/mohammed_almakki/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,123 @@
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <regex>
+#include <unordered_map>
+#include <cstring>
+#include <iomanip>
+
+std::string readFile(const std::string &filepath)
+{
+    std::ifstream inputStream(filepath);
+
+    if (inputStream)
+    {
+        std::ostringstream ss;
+        ss << inputStream.rdbuf();
+        return ss.str();
+    }
+    else
+    {
+        throw std::invalid_argument("Invalid path. Please make sure the file path is correct and the file exists.");
+    }
+}
+
+std::vector<std::string> tokenizeFile(const std::string &file)
+{
+    std::regex re("[-\\s]");
+    std::sregex_token_iterator first{file.begin(), file.end(), re, -1}, last;
+    std::vector<std::string> tokens{first, last};
+    return tokens;
+}
+
+std::unordered_map<std::string, uint32_t> countWords(const std::vector<std::string> words)
+{
+    std::unordered_map<std::string, uint32_t> wordsCounts;
+    for (auto word : words)
+    {
+        std::transform(
+            word.begin(), word.end(), word.begin(),
+            [](unsigned char c)
+            { return std::tolower(c); });
+        word.erase(std::remove_if(word.begin(), word.end(), ispunct), word.end());
```

Comment:
The task specifies only needing to consider `.,?'"!():` you might have considered using regex to strip these away and replacing them with a whitespace at the same time as doing the hyphen.

Reply:
As I understood, the comparison should be punctation insensitive. So for example, "what's" should be equal to "whats" but replacing it with whitespace will result in the word not being included (not longer than 4). So in this line, I just remove every punctuation marks.

Reply:
In that case I think you may be right